routers/web/auth/webauthn.go

```diff
@@ -0,0 +1,158 @@
+		return
+	}
+
+	//Ensure user is in a 2FA session.
+	if ctx.Session.Get("twofaUid") == nil {
+		ctx.ServerError("UserSignIn", errors.New("not in WebAuthn session"))
+		return
+	}
+
+	ctx.HTML(200, tplWebAuthn)
+}
+
+// WebAuthnLoginAssertion submits a WebAuthn challenge to the browser
+func WebAuthnLoginAssertion(ctx *context.Context) {
+	// Ensure user is in a WebAuthn session.
+	idSess := ctx.Session.Get("twofaUid")
+	if idSess == nil {
+		ctx.ServerError("UserSignIn", errors.New("not in WebAuthn session"))
+		return
+	}
+
+	user, err := user_model.GetUserByID(idSess.(int64))
+	if err != nil {
+		ctx.ServerError("UserSignIn", err)
+		return
+	}
+
+	creds, err := auth.WebAuthnCredentials(user.ID)
+	if err != nil {
+		ctx.ServerError("UserSignIn", err)
+		return
+	}
+	if len(creds) == 0 {
+		ctx.ServerError("UserSignIn", errors.New("no device registered"))
+		return
+	}
+
+	userVerification := ctx.FormString("user_ver")
+	txAuthSimpleExtension := ctx.FormString("tx_auth_simple_extension")
+	testExtension := protocol.AuthenticationExtensions(map[string]interface{}{"txAuthSimple": txAuthSimpleExtension})
+
+	assertion, sessionData, err := wa.WebAuthn.BeginLogin(
+		(*wa.User)(user),
+		webauthn.WithUserVerification(protocol.UserVerificationRequirement(userVerification)),
+		webauthn.WithAssertionExtensions(testExtension),
+	)
+	if err != nil {
+		ctx.ServerError("webauthn.BeginLogin", err)
+		return
+	}
+
+	if err := ctx.Session.Set("webauthnAssertion", sessionData); err != nil {
+		ctx.ServerError("Session.Set", err)
+		return
+	}
+	ctx.JSON(http.StatusOK, assertion)
+}
+
+// WebAuthnLoginAssertionPost validates the signature and logs the user in
+func WebAuthnLoginAssertionPost(ctx *context.Context) {
+	idSess := ctx.Session.Get("twofaUid")
+	sessionData := ctx.Session.Get("webauthnAssertion")
+	if sessionData == nil || idSess == nil {
+		ctx.ServerError("UserSignIn", errors.New("not in WebAuthn session"))
+		return
+	}
+	defer func() {
+		_ = ctx.Session.Delete("webauthnAssertion")
+	}()
+	user, err := user_model.GetUserByID(idSess.(int64))
+	if err != nil {
+		ctx.ServerError("UserSignIn", err)
+		return
+	}
+
+	log.Trace("Finishing webauthn authentication with user: %s\n", user.Name)
+
+	// With the session data retrieved, we need to call webauthn.FinishLogin to
+	// verify the signed challenge. This returns the webauthn.Credential that
+	// was used to authenticate.
+	cred, err := wa.WebAuthn.FinishLogin((*wa.User)(user), *sessionData.(*webauthn.SessionData), ctx.Req)
+	if err != nil {
+		ctx.ServerError("FinishLogin", err)
+		return
+	}
+
+	// At this point, we've confirmed the correct authenticator has been
+	// provided and it passed the challenge we gave it. We now need to make
+	// sure that the sign counter is higher than what we have stored to help
+	// give assurance that this credential wasn't cloned.
+	if cred.Authenticator.CloneWarning {
+		log.Error("credential appears to be cloned: %s", err)
+		ctx.Status(http.StatusForbidden)
```

Comment:
We need to log a failed authentication attempt and this should send an error back that webAuthnError in the javascript can handle.